### PR TITLE
LSP: fix "call hierarchy" across files

### PR DESCRIPTION
**Pull request by DanilaFe**
In the following program, trying to get the call hierarchy for `writeln` did not work on `main`.

```Chapel
writeln("Hello, world!");
```

There were a few issues with this. The first issue is that the the "index" that was fed through a `CallHierarchyItem` to LSP and back to `CLS` was an index into a single file's list of instantiations. Thus, if the function was instantiated in one place, but defined in another, the index would not be valid (or worse, be incorrect). This PR changes this by changing the way typed signatures are mapped to JSON-compatible identifiers; they are instead registered with a `ContextContainer` object, getting a unique ID. The unique ID is then used as JSON data, and can be used to access the same `TypedSignature` object when the JSON is being deserialized.

This, however, was not enough. The next problem was that we currently create a new `ContextContainer` per file (to handle the case in which multiple files with the same name / module name exist). As a result of this, `writeln` that is called in module `A` and defined in module `B` will have its instantiation computed using `A`'s context, but its AST and other information will use `B`'s context. This is troublesome. This PR adjusts the situation by changing the `file -> FileInfo` mapping to be a `(file, context id) -> FileInfo` mapping. This way, the same file can have several contexts (e.g., if a standard library file is loaded from two distinct files, but both named `A.chpl`). By default (e.g., when opening a file), the `(file, "None")` key is used (since no particular context was requested). However, if a file A needs to show an instantiation (via file info) from file B, we create a new mapping `(fileB, fileA.context.id) -> fileInfoB`, which is created from the same context. This ensures that there is a file information object for file B with the same context as that of file A.

When a call hierarchy item is constructed, it now specifies what context was used to create it. Thus, if other CallHierarchyItems point to other files, these other files are opened with the same context. To communicate the context over JSON, I employ the same approach as I did with typed signatures (register a context object with unique ID, use the ID to marshal the context to and from JSON).

~~To fix the case of `writeln` in particular, I had to relax an assertion from the `_owned`/`_shared_` resolution PR https://github.com/chapel-lang/chapel/pull/25617. This led to some simplified code in any case, which is great. I believe @arezaii is fine with this change.~~ This was separately merged in https://github.com/chapel-lang/chapel/pull/26078.

A bug that I'm surprised we didn't find prior to this, which is fixed in this PR, is that vector unmarshalling for Python interop allocates a vector of the Python list's size, then pushes _additional elements_ instead of setting the already-allocated locations. This ended up e.g., pushing an empty string for each "real" string a user passed in a list. This led to some odd interactions with Dyno's module searching (an empty string as an input file led to "." being added to the module search path, which we do not want).

Reviewed by @jabraham17 -- thanks!

## Testing
- [x] LSP tests
- [x] new call hierarchy tests
- [x] dyno tests

### CODE REVIEW

File: frontend/lib/resolution/Resolver.cpp [hunk 1]
```diff
@@ -2511,17 +2511,13 @@ QualifiedType Resolver::typeForId(const ID& id, bool localGenericToUnknown) {
...
-        auto borrowed = nct->withDecorator(context, nct->decorator().toBorrowed());
-        return QualifiedType(intent, borrowed);
-      } else if (auto comprt = rt->getCompositeType()) {
+      if (auto ct = rt->toClassType()) {
+        if (auto mr = ct->managerRecordType(context)) {
+          rt = mr;
+        }
+      }
+
+      if (auto comprt = rt->getCompositeType()) {
```
Comment (jabraham17) [inline frontend/lib/resolution/Resolver.cpp:2520]: You should get someone on the dyno team to confirm this part as well

File: tools/chpl-language-server/src/chpl-language-server.py [hunk 2]
```diff
@@ -499,6 +499,8 @@ def __init__(self, file: str, config: Optional["WorkspaceConfig"]):
         self.context: chapel.Context = chapel.Context()
         self.file_infos: List["FileInfo"] = []
         self.global_uses: Dict[str, List[References]] = defaultdict(list)
+        self.instantiation_ids: Dict[chapel.TypedSignature, str] = {}
```
Comment (jabraham17) [inline tools/chpl-language-server/src/chpl-language-server.py:502]: I am surprised that TypedSignature can be used as the key for a dict, I guess it provides a hash method?

File: tools/chpl-language-server/src/chpl-language-server.py [hunk 3]
```diff
@@ -1128,19 +1164,34 @@ def get_file_info(
         creating one if it doesn't exist. If do_update is set to True,
         then the FileInfo's index is rebuilt even if it has already been
         computed. This is useful if the underlying file has changed.
+
+        Most of the tiem, we will create a new context for a given URI. When
```
Comment (jabraham17) [inline tools/chpl-language-server/src/chpl-language-server.py:1168]: nit, typo

File: tools/chpl-language-server/src/chpl-language-server.py [hunk 4]
```diff
@@ -1433,16 +1485,17 @@ def unpack_call_hierarchy_item(
...
+            or not isinstance(item.data[1], Optional[str])
+            or not isinstance(item.data[2], Optional[str])
         ):
             self.show_message(
                 "Call hierarchy item contains missing or invalid additional data",
                 MessageType.Error,
             )
             return None
-        uid, idx = item.data
+        uid, idx, ctx = item.data
```
Comment (jabraham17) [inline tools/chpl-language-server/src/chpl-language-server.py:1496]: nit, rename `idx` to `inst_id`

**Review by jabraham17 — APPROVED** [review]
looks good, just a few nits

File: frontend/lib/resolution/Resolver.cpp (hunk 1 above)
Comment (DanilaFe) [reply frontend/lib/resolution/Resolver.cpp:2520]: I did, I checked with Ahmad.

File: tools/chpl-language-server/src/chpl-language-server.py (hunk 2 above)
Comment (DanilaFe) [reply tools/chpl-language-server/src/chpl-language-server.py:502]: I did specifically add one in a previous PR, it just hashes the PoiScope and TypedFnSignature pointers.

File: tools/chpl-language-server/test/call_hierarchy.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
+"""
+Tests basic functionality, including autocompletion, go-to-definition, hover,
```
Comment (jabraham17) [inline tools/chpl-language-server/test/call_hierarchy.py:2]: nit, wrong file comment

**Review by jabraham17 — APPROVED** [review]

**Pull request merged**